fuel/utils/cache.py

```diff
@@ -0,0 +1,264 @@
+            log.warning(common_msg +
+                        "File {} not cached: Not enough free space"
+                        .format(remote_name))
+            release_writelock()
+            return filename
+
+        # There is enough space; make a local copy of the file
+        copy_from_server_to_local(dataset_remote_dir, dataset_local_dir,
+                                  remote_name, local_name)
+        log.info(common_msg + "File {} has been locally cached to {}"
+                 .format(remote_name, local_name))
+    elif os.path.getmtime(remote_name) > os.path.getmtime(local_name):
+        remote_modified_time = time.strftime(
+            '%Y-%m-%d %H:%M:%S',
+            time.localtime(os.path.getmtime(remote_name)))
+        local_modified_time = time.strftime(
+            '%Y-%m-%d %H:%M:%S',
+            time.localtime(os.path.getmtime(local_name)))
+        log.warning(common_msg +
+                    "File {} in cache will not be used: The remote file "
+                    "(modified {}) is newer than the locally cached file "
+                    "{} (modified {})."
+                    .format(remote_name, remote_modified_time,
+                            local_name, local_modified_time))
+        release_writelock()
+        return filename
+    elif os.path.getsize(local_name) != os.path.getsize(remote_name):
+        log.warning(common_msg +
+                    "File {} not cached: The remote file ({} bytes) is of "
+                    "a different size than the locally cached file {} "
+                    "({} bytes). The local cache might be corrupt."
+                    .format(remote_name, os.path.getsize(remote_name),
+                            local_name, os.path.getsize(local_name)))
+        release_writelock()
+        return filename
+    elif not os.access(local_name, os.R_OK):
+        log.warning(common_msg +
+                    "File {} in cache isn't readable. We will use the"
+                    " remote version. Manually fix the permission."
+                    .format(local_name))
+        release_writelock()
+        return filename
+    else:
+        log.debug("File {} has previously been locally cached to {}"
+                  .format(remote_name, local_name))
+
+    # Obtain a readlock on the downloaded file before releasing the
+    # writelock. This is to prevent having a moment where there is no
+    # lock on this file which could give the impression that it is
+    # unused and therefore safe to delete.
+    get_readlock(pid, local_name)
+    release_writelock()
+
+    return local_name
+
+
+def copy_from_server_to_local(dataset_remote_dir, dataset_local_dir,
+                              remote_fname, local_fname):
+    """Copies a remote file locally.
+
+    Parameters
+    ----------
+    remote_fname : str
+        Remote file to copy
+    local_fname : str
+        Path and name of the local copy to be made of the remote file.
+
+    """
+
```

Comment:
Can we add `log.debug` here, so that impatient users could see that their files are being copied?